File: bench/jl/bench_loading.jl
#!/usr/bin/env julia
# CSV parse benchmark — base (readdlm) vs pkg (CSV.jl)

function loading_base(path)
    lines = readlines(path)
    data = lines[4:end]
    n = length(data)
    close = Vector{Float64}(undef, n)
    high  = Vector{Float64}(undef, n)
    low   = Vector{Float64}(undef, n)
    open_ = Vector{Float64}(undef, n)
    vol   = Vector{Float64}(undef, n)
    for i in 1:n
        parts = split(data[i], ',')
        close[i] = parse(Float64, parts[2])
        high[i]  = parse(Float64, parts[3])
        low[i]   = parse(Float64, parts[4])
        open_[i] = parse(Float64, parts[5])
        vol[i]   = parse(Float64, parts[6])
    end
    # Checksum to prevent dead-code elimination
    println(repr(sum(close) + sum(high) + sum(low) + sum(open_) + sum(vol)))
end

using CSV, DataFrames

function loading_pkg(path)
    df = CSV.read(path, DataFrame; header=3, skipto=4)
    c = df[!, 2]
    h = df[!, 3]
    l = df[!, 4]
    o = df[!, 5]
    v = df[!, 6]
    println(repr(sum(c) + sum(h) + sum(l) + sum(o) + sum(v)))
end

function main()
    csv_path = ARGS[1]
    mode = length(ARGS) >= 3 && ARGS[2] == "--mode" ? ARGS[3] : "base"

    if mode == "base"
        loading_base(csv_path)
    else
        loading_pkg(csv_path)
    end
end

main()

File: bench/py/bench_loading.py
#!/usr/bin/env python3
"""CSV parse only benchmark — pandas vs numpy (manual)."""
import sys
import argparse
import numpy as np
import pandas as pd


def loading_pandas(path):
    """Load via pd.read_csv."""
    df = pd.read_csv(
        path, skiprows=3, header=None,
        names=["date", "close", "high", "low", "open", "volume"],
    )
    checksum = (df["close"].sum() + df["high"].sum() + df["low"].sum()
                + df["open"].sum() + df["volume"].sum())
    print(repr(float(checksum)))


def loading_numpy(path):
    """Load via manual line parsing (matches BQN approach)."""
    with open(path) as f:
        lines = f.readlines()[3:]  # skip 3 header lines
    n = len(lines)
    close = np.empty(n)
    high = np.empty(n)
    low = np.empty(n)
    open_ = np.empty(n)
    vol = np.empty(n)
    for i, line in enumerate(lines):
        parts = line.strip().split(",")
        close[i] = float(parts[1])
        high[i] = float(parts[2])
        low[i] = float(parts[3])
        open_[i] = float(parts[4])
        vol[i] = float(parts[5])
    checksum = close.sum() + high.sum() + low.sum() + open_.sum() + vol.sum()
    print(repr(float(checksum)))


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("csv", help="Path to OHLCV CSV")
    parser.add_argument("--mode", choices=["pandas", "numpy"], default="pandas")
    args = parser.parse_args()
    if args.mode == "pandas":
        loading_pandas(args.csv)
    else:
        loading_numpy(args.csv)
